src/index/src/inverted_index/search/fst_apply/intersection_apply.rs

```diff
@@ -32,7 +35,41 @@ pub struct IntersectionFstApplier {
     ranges: Vec<Range>,
 
     /// A list of `Dfa` compiled from regular expression patterns.
-    dfas: Vec<DFA<Vec<u32>>>,
+    dfas: Vec<DfaFstAutomaton>,
+}
+
+struct DfaFstAutomaton(DFA<Vec<u32>>);
+
+impl fst::Automaton for DfaFstAutomaton {
+    type State = StateID;
+
+    #[inline]
+    fn start(&self) -> Self::State {
+        let config = Config::new();
```

Comment:
> Seems some internal impl issues exist; the following cases failed:
> 
> ```
> index inverted_index::search::fst_apply::intersection_apply::tests::test_intersection_fst_applier_with_valid_pattern
> mito2 sst::index::creator::tests::test_create_and_query_regex
> ```

try `Config::new().anchored(Anchored::Yes)`?

Reply:
Tried. Doesn't seem help :<

Reply:
not familiar with regex. I tweaked it a bit and couldn't get it to work...

Reply:
The regex should be correct but the state machine may be incompatible. It's a bit hard to debug but if I can draw the DFA ..

Perhaps @BurntSushi can be clear if we wrap the struct wrongly. The issue seems related to backward search. That is, `a.*` consume all the pattern, and the trailing `c` fail to match anything. Typically, the search machine should move one step backward and try to match c.

Reply:
No. Pattern `abc` and `a.c` doesn't match also.

Reply:
I'm reading the docs to see how to use `dfa::automata::Automaton` ...

Reply:
```
        let next_state = self.0.next_eoi_state(*state);
        self.0.is_match_state(next_state)
```

helps on unanchored case. But I'm unsure if always all `next_eoi_state` for each `is_match` call efficient.

`\\w` and `\\d` still yet to work. Debugging ...

Reply:
add a guard to prevent over moving the state https://github.com/GreptimeTeam/greptimedb/pull/3575/commits/8b9d6f88494dd6a21891b954894d1b359f54fda1

UT passed now. I'm unsure if it's correct logically, though. Seems fst and regex-automata has different design for handling state.